crates/polars-stream/src/nodes/io_sources/batch.rs

```diff
@@ -1,118 +1,188 @@
+//! Reads batches from a `dyn Fn`
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
 use polars_core::frame::DataFrame;
 use polars_core::schema::SchemaRef;
 use polars_error::{PolarsResult, polars_err};
+use polars_expr::state::ExecutionState;
+use polars_utils::IdxSize;
 use polars_utils::pl_str::PlSmallStr;
-use polars_utils::{IdxSize, format_pl_smallstr};
-use tokio::sync::oneshot;
 
-use super::{
-    JoinHandle, Morsel, MorselSeq, SourceNode, SourceOutput, StreamingExecutionState, TaskPriority,
+use crate::async_executor::{JoinHandle, TaskPriority, spawn};
+use crate::execute::StreamingExecutionState;
+use crate::morsel::{Morsel, MorselSeq, SourceToken};
+use crate::nodes::io_sources::multi_file_reader::reader_interface::output::{
+    FileReaderOutputRecv, FileReaderOutputSend,
+};
+use crate::nodes::io_sources::multi_file_reader::reader_interface::{
+    BeginReadArgs, FileReader, FileReaderCallbacks,
 };
-use crate::async_executor::spawn;
-use crate::async_primitives::connector::Receiver;
-use crate::async_primitives::wait_group::WaitGroup;
-use crate::morsel::SourceToken;
 
-type GetBatchFn =
+pub mod builder {
+
+    use std::sync::{Arc, Mutex};
+
+    use polars_utils::pl_str::PlSmallStr;
+
+    use super::BatchFnReader;
+    use crate::nodes::io_sources::multi_file_reader::reader_interface::FileReader;
+    use crate::nodes::io_sources::multi_file_reader::reader_interface::builder::FileReaderBuilder;
+    use crate::nodes::io_sources::multi_file_reader::reader_interface::capabilities::ReaderCapabilities;
+
+    pub struct BatchFnReaderBuilder {
+        pub name: PlSmallStr,
+        pub reader: Mutex<Option<BatchFnReader>>,
+    }
+
+    impl FileReaderBuilder for BatchFnReaderBuilder {
+        fn reader_name(&self) -> &str {
+            &self.name
+        }
+
+        fn reader_capabilities(&self) -> ReaderCapabilities {
+            ReaderCapabilities::empty()
```

Comment:
The IO plugin sources currently assume that they have all capabilities. I am going to add flags to all customizing that.